src/main/java/com/imsweb/mph/MphUtils.java

```diff
@@ -4,40 +4,69 @@
+import com.imsweb.mph.mpgroups.Mp1998HematopoieticGroup;
+import com.imsweb.mph.mpgroups.Mp2001HematopoieticGroup;
+import com.imsweb.mph.mpgroups.Mp2004BenignBrainGroup;
+import com.imsweb.mph.mpgroups.Mp2004SolidMalignantGroup;
+import com.imsweb.mph.mpgroups.Mp2007BenignBrainGroup;
+import com.imsweb.mph.mpgroups.Mp2007BreastGroup;
+import com.imsweb.mph.mpgroups.Mp2007ColonGroup;
+import com.imsweb.mph.mpgroups.Mp2007HeadAndNeckGroup;
+import com.imsweb.mph.mpgroups.Mp2007KidneyGroup;
+import com.imsweb.mph.mpgroups.Mp2007LungGroup;
+import com.imsweb.mph.mpgroups.Mp2007MalignantBrainGroup;
+import com.imsweb.mph.mpgroups.Mp2007MelanomaGroup;
+import com.imsweb.mph.mpgroups.Mp2007OtherSitesGroup;
+import com.imsweb.mph.mpgroups.Mp2007UrinaryGroup;
+import com.imsweb.mph.mpgroups.Mp2010HematopoieticGroup;
 
 /**
- * This class is used to determine whether two tumors are single or multiple primaries. More information can be found here:
- * <a href="http://www.seer.cancer.gov/tools/mphrules">http://www.seer.cancer.gov/tools/mphrules</a>
+ * This class is used to determine single versus multiple primaries. More information can be found on the following websites:
+ * <a href="http://seer.cancer.gov/archive/manuals/codeman.pdf">SEER 1998 multiple primary rules for hematopoietic cancer</a>
  * <br/><br/>
- * This Java implementation is based on the the PDF documentation provided on the above website.
+ * <a href="http://training.seer.cancer.gov/hemat/">SEER 2001 multiple primary rules for hematopoietic cancer</a>
+ * <br/><br/>
+ * <a href="http://seer.cancer.gov/tools/heme/Hematopoietic_Instructions_and_Rules.pdf">SEER multiple primary rules for hematopoietic cancer for cases diagnosed in 2010 and later</a>
+ * <br/><br/>
+ * <a href="http://seer.cancer.gov/archive/manuals/2004Revision1/SPM_2004_maindoc.r1.pdf">SEER 2004 multiple primary rules</a>
+ * <br/><br/>
+ * <a href="http://www.seer.cancer.gov/tools/mphrules">SEER 2007 multiple primary rules</a>
+ * <br/><br/>
+ * This Java implementation is based on the the documentation provided on the above websites.
  * Created in December 2013 by Sewbesew Bekele
  */
 public class MphUtils {
 
-    public static final String PROP_PRIMARY_SITE = "primarySite";
-    public static final String PROP_HISTOLOGY_ICDO3 = "histologyIcdO3";
-    public static final String PROP_BEHAVIOR_ICDO3 = "behaviorIcdO3";
-    public static final String PROP_LATERALITY = "laterality";
-    public static final String PROP_DX_YEAR = "dateOfDiagnosisYear";
-    public static final String PROP_DX_MONTH = "dateOfDiagnosisMonth";
-    public static final String PROP_DX_DAY = "dateOfDiagnosisDay";
+    private static MphUtils _UTILS = null;
+    private HematoDbUtilsProvider _provider = null;
+    private List<MphGroup> _groups = new ArrayList<>();
+    private static final Object _INIT_LOCK = new Object();
+
+    public static void initialize(HematoDbUtilsProvider provider) {
+        synchronized (_INIT_LOCK) {
```

Comment:
Why not just synchronize the initialize method and return early in that method if you detect the initialization has already been done. I think that would cover all the cases and that would be simpler than using an explicit lock...